wallets/core/src/hub/namespaces/namespace.ts

```diff
@@ -0,0 +1,443 @@
+   * ```ts
+   *  const actions = new Map();
+   *  actions.set('connect', connectAction);
+   *
+   *  const ns = new Namespace(NAMESPACE, PROVIDER_ID, {
+   *    actions: actions,
+   *  });
+   *
+   *  ns.run("action");
+   * ```
+   */
+  public run<K extends keyof T>(
+    actionName: K,
+    ...args: any[]
+  ): unknown | Promise<unknown> {
+    const action = this.#actions.get(actionName);
+    if (!action) {
+      throw new Error(ACTION_NOT_FOUND_ERROR(actionName.toString()));
+    }
+
+    /*
+     * Action can be both, sync or async. To simplify the process we can not make `sync` mode to async
+     * Since every user's sync action will be an async function and affect what user expect,
+     * it makes all the actions async and it doesn't match with Namespace interface (e.g. EvmActions)
+     *
+     * To avoid this issue and also not duplicating code, I broke the process into smaller methods
+     * and two main methods to run actions: tryRunAsyncAction & tryRunAction.
+     */
+    const result = isAsync(action)
+      ? this.#tryRunAsyncAction(actionName, args)
+      : this.#tryRunAction(actionName, args);
+
+    return result;
+  }
+
+  #tryRunAction<K extends keyof T>(actionName: K, params: any[]): unknown {
+    this.#tryRunBeforeHooks(actionName);
+
+    const action = this.#actions.get(actionName);
+    if (!action) {
+      throw new Error(ACTION_NOT_FOUND_ERROR(actionName.toString()));
+    }
+
+    const context = this.#context();
+
+    let result;
+    try {
+      result = action(context, ...params);
+      result = this.#tryRunAndOperators(actionName, result);
+    } catch (e) {
+      result = this.#tryRunOrOperators(actionName, e);
+    } finally {
+      this.#tryRunAfterHooks(actionName);
+    }
+
+    return result;
+  }
+
+  async #tryRunAsyncAction<K extends keyof T>(
+    actionName: K,
+    params: any[]
+  ): Promise<unknown> {
+    this.#tryRunBeforeHooks(actionName);
+
+    const action = this.#actions.get(actionName);
+    if (!action) {
+      throw new Error(ACTION_NOT_FOUND_ERROR(actionName.toString()));
+    }
+
+    const context = this.#context();
+    return await action(context, ...params)
+      .then((result: unknown) => this.#tryRunAndOperators(actionName, result))
+      .catch((e: unknown) => this.#tryRunOrOperators(actionName, e))
+      .finally(() => this.#tryRunAfterHooks(actionName));
+  }
+
+  #tryRunAfterHooks<K extends keyof T>(actionName: K) {
+    const afterActions = this.#afterHooks.get(actionName);
+
+    if (afterActions) {
+      afterActions.forEach((afterAction) => {
+        const context = afterAction.options?.context || this.#context();
+        afterAction.hook(context);
+      });
+    }
+  }
+
+  #tryRunBeforeHooks<K extends keyof T>(actionName: K): void {
+    const beforeActions = this.#beforeHooks.get(actionName);
+    if (beforeActions) {
+      beforeActions.forEach((beforeAction) => {
+        const context = beforeAction.options?.context || this.#context();
```

Comment:
I'm not sure what do you mean by `state`. But if you need an explanation on this line, this will use `Namespace` context by default, but in some cases it can be set `Provider`. Which means a Provider now can define a context for its naemspaces.